#17422 improved

haampie opened the pull request:
This includes #17422, but fixes a list of issues with it:

- no more strict dependency on cmake 3.5, but simply allow 3 and above (required a patch)
- fixes a problem where `rocclr` was built outside of the staging area and then reused in dependents
- fixes compilation of HsakmtRoct which had missing dependencies and flags that made the build fail
- fixes RPATH issues (required a patch for `hsa-rocr-dev` which is also submitted upstream)
- adds a variant for the amd gpu target for rocfft, because hipcc tries to autodetect the gpu target, for which it needs a binary which it assumes to be in a specific folder which we can't have in spack
- Fixes the tests

srekolam committed:
Spack recipes for ROCm Stage 1 Build components

Commit:
fix flake8 errors

srekolam committed:
fixes for flake8 errors

haampie committed:
Use cmake 3.x instead of a fixed version

haampie committed:
Add a patch for cmake 3.x suport

haampie committed:
Fix rpath issue where hsa-rocr-dev does not allow it to be filled in by spack

haampie committed:
Remove inherited cmake args from comgr

haampie committed:
Make hsakmt-roct compile: no -Werror because with const cast in numa, and actually add the numa dependency

haampie committed:
Remove redundant cmake args which is inherited

haampie committed:
Fix some dependencies

haampie committed:
Fix some python 2.x compatibilities

haampie committed:
Add amd gpu targets to rocfft

haampie committed:
Make comgr a link dep of rocm-dbgapi and remove redundant cmake args

haampie committed:
Remove redundant cmake args

haampie committed:
Remove more redundant cmake args

haampie committed:
Final redundant args

haampie committed:
Remove random variable

haampie committed:
Use installed rocclr instead of nonexisting directory

haampie committed:
Don't build outside the staging folder

haampie committed:
Deploy some missing cmake target file

haampie committed:
Formatting

haampie committed:
Fix target list

haampie committed:
Properly handle the rocclr dependency

haampie committed:
Formatting

haampie committed:
Fix vermin test

haampie committed:
Make all 3.5.0 package depend exactly on eachother

haampie commented:
I think this is mostly correct now. It mimics the AUR more or less. There are some issues upstream still, such as https://github.com/RadeonOpenCompute/rocminfo/issues/9.

haampie committed:
Add a few missing link dependencies

haampie committed:
Fix flake8

haampie committed:
Remove some other redundant flags

srekolam commented:
@haampie , We did some additional changes on top of pull request (https://github.com/srekolam/spack/pull/1) . I am thinking we will validate everything  our side including yours and then update PR17422. 
@adamjstewart , I need some inputs on the best way forward.

haampie commented:
@srekolam I think it's easiest if you would just merge https://github.com/srekolam/spack/pull/1, and then add your changes on top of that, so that we can continue with the review in #17422 and close this PR.

This PR has everything rebased on the spack develop branch, so it's not easy to merge back into your PR. https://github.com/srekolam/spack/pull/1 has all commits cherry-picked and should be easier to work with.

aweits commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -0,0 +1,34 @@
+                '-DCMAKE_VERBOSE_MAKEFILE=1',
+                '-DLLVM_ENABLE_PROJECTS=clang;lld;clang-tools-extra;compiler-rt',
+                '-DLLVM_ENABLE_ASSERTIONS=1'
+               ]
```
We're going to need this here as well. tagging @naromero77 to get some extra eyes on this

```
        if self.compiler.name == "gcc":
            gcc_prefix = ancestor(self.compiler.cc, 2)
            args.append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)
```

haampie replied:
Sure, let me add it. What does it do exactly?

Edit: I see, it's also part of the normal llvm package

naromero77 replied:
Yes, that seems necessary on some Linux OS while others don't seem to require it. Safer to just always specify it.

haampie committed:
Add gcc install prefix for gcc builds of llvm-amdgpu

srekolam committed:
review changes for the spack recipes

haampie committed:
Do not hard-code versions

haampie committed:
Fix atmi install

- no more relative rpaths outside of install directory (required patch)
- fix build -> link dependencies
- remove unused build dependency

haampie committed:
Fix flake8 errors

haampie committed:
Remove unused variable and make things python 2.x compatible

haampie committed:
Fix flake8

haampie commented:
@aweits @naromero77: your suggestion `append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)` breaks clang++ for me :(

```
harmen-desktop tmp $ cat main.cc 
#include <iostream>

int main() {
    std::cout << "hello\n";
}

harmen-desktop tmp $ # with the suggestion
harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-4pxgxhxmrdc6sd7fgz5c5xndmtj6ikqy/bin/clang++ main.cc 
main.cc:1:10: fatal error: 'iostream' file not found
#include <iostream>
         ^~~~~~~~~~
1 error generated.

harmen-desktop tmp $ # without the suggestion
harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-3mqgueaylvsdnzgenyd4fnqtonhji7vs/bin/clang++ main.cc 
harmen-desktop tmp $ ./a.out 
hello
```

Apparently it gets `GCC_INSTALL_PREFIX=/`.

Edit: this is because `spack compiler find` sets the paths to `/bin/g++` on Ubuntu 20.04. :cry:. I guess it works if I change that to `/usr/bin/g++`

I'm gonna leave it in anyways, see https://github.com/spack/spack/issues/17932, people should just change their system compiler paths then.

aweits commented:
> @aweits @naromero77: your suggestion `append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)` breaks clang++ for me :(

I'm curious, what does the output from 'clang++ -v' give you in the broken case?

haampie committed:
Move compiler config from rocfft -> hipcc

haampie committed:
Remove redundant dependency on fftw-api

haampie committed:
Make spfft build with rocm :rocket:

haampie committed:
Remove redundant import

naromero77 commented:
> @aweits @naromero77: your suggestion `append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)` breaks clang++ for me :(
> 
> ```
> harmen-desktop tmp $ cat main.cc 
> #include <iostream>
> 
> int main() {
>     std::cout << "hello\n";
> }
> 
> harmen-desktop tmp $ # with the suggestion
> harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-4pxgxhxmrdc6sd7fgz5c5xndmtj6ikqy/bin/clang++ main.cc 
> main.cc:1:10: fatal error: 'iostream' file not found
> #include <iostream>
>          ^~~~~~~~~~
> 1 error generated.
> 
> harmen-desktop tmp $ # without the suggestion
> harmen-desktop tmp $ /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-amdgpu-3.5.0-3mqgueaylvsdnzgenyd4fnqtonhji7vs/bin/clang++ main.cc 
> harmen-desktop tmp $ ./a.out 
> hello
> ```
> 
> Apparently it gets `GCC_INSTALL_PREFIX=/`.
> 
> Edit: this is because `spack compiler find` sets the paths to `/bin/g++` on Ubuntu 20.04. 😢. I guess it works if I change that to `/usr/bin/g++`
> 
> I'm gonna leave it in anyways, see #17932, people should just change their system compiler paths then.

Sure -- sorry, that is annoying

haampie committed:
Avoid hitting the ROCM_PATH variable altogether with a patch; also just fill in all variables

haampie committed:
Add missing deps z3, zlib and ncurses+termlib to llvm-amdgpu

haampie committed:
Fix perl shebang and add dep

adamjstewart commented:
Can we close this PR and focus reviews on #17422?

haampie commented:
I would be in favor of that too, but #17422 is missing some commits (https://github.com/srekolam/spack/pull/2).

adamjstewart commented:
@srekolam would it be possible to give @haampie merge privileges to your fork so you two can collaborate on a single branch?

haampie committed:
Add rocBLAS

haampie committed:
Fix typo and patch HIP_CLANG_ROOT detection in hip's cmake files

haampie committed:
Finish rocBLAS support and add it to the sirius package

srekolam commented:
@haampie , thanks for your inputs..we have the recipes for rocblas and other math libs too.. we plan to raise the PR once the current 2 are merged into the spack.

srekolam commented:
@haampie , i see some build failures in HIP after pulling your 2&3 , I have reverted back to commit id 9c6f763814813541633d147fa92497eb11073a68. I will try to cherry pick  others if needed from 2&3

srekolam commented:
After merging https://github.com/srekolam/spack/pull/2 , i uninstalled everything and rebuilt hsakmt-roct, hsa-rocr-dev,llvm-amdgpu . I then pulled https://github.com/srekolam/spack/pull/3 and tried to build hip.. Not sure if this is causing the issue.. I have actually been able to build and do some tests for https://github.com/srekolam/spack/pull/1..

adamjstewart commented:
Since all of these changes have been merged into #17422, I'm going to close this so we can focus the review on that PR.
